DX-1154 refactor interfaces with SignableToken

zatlite opened the pull request:
# Summary
Refactor interfaces with `SignableToken` to use `AnyToken` union type

[DX-1154]

# Why the changes
`SignableToken` is currently not properly typed because API gen code uses OAS 2.0 spec which is missing support for `OneOf` type.
We implement our own interfaces to be exposed to the SDK user at the same time we flatten the data structure to remove the `data` object.

# Things worth calling out
<!--- Give useful tips/gotchas/trade-offs made to the reviewers. -->


[DX-1154]: https://immutable.atlassian.net/browse/DX-1154?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

zatlite committed:
fix: typo in function name

zatlite committed:
refactor: interfaces with SignableToken to use AnyToken union type

[DX-1154]

zatlite committed:
refactor: using `extends Omit`

zatlite committed:
refactor: minor changes

Kastet commented:
@zatlite can you show example usages to be able to see how compares with the current interface

zatlite commented:
> @zatlite can you show example usages to be able to see how compares with the current interface

examples: https://github.com/immutable/imx-core-sdk-tester/pull/2

zatlite committed:
refactor: as per review comment

zatlite committed:
refactor: remove sender from transfer interfaces

sender eth address can be filled in using the wallet connection

Kastet commented on src/types/custom-interfaces.ts:
```diff
@@ -0,0 +1,51 @@
+}
+
+export interface UnsignedTransferRequest
+  extends Omit<GetSignableTransferRequestV1, 'token' | 'sender'> {
```
Let's not use this pattern of extending API types and instead just type out the whole interface. In IDE you need to click through many level deep to see the interface.

zatlite replied:
[502c68a](https://github.com/immutable/imx-core-sdk/pull/149/commits/502c68a0ab770b0463b90a249b37de5ae524fb77)

Kastet commented on src/types/custom-interfaces.ts:
```diff
@@ -0,0 +1,51 @@
+  token: AnyToken;
+}
+
+export interface NftTransferDetails extends Omit<ERC721Token, 'type'> {
```
Same comment here, let's just type out the whole interface

zatlite replied:
[502c68a](https://github.com/immutable/imx-core-sdk/pull/149/commits/502c68a0ab770b0463b90a249b37de5ae524fb77)

Kastet commented on src/ImmutableX.ts:
```diff
@@ -1,11 +1,14 @@
   TokenDeposit,
   TokenWithdrawal,
-  EthSigner,
+  UnsignedMintRequest,
+  UnsignedOrderRequest,
+  UnsignedTransferRequest,
```
Do these request types need "Unsigned" work in them? For the user of the SDK these are just requests and SDK internally handles the signing. Suggest removing the "Unsigned"

zatlite replied:
Would clash with generated api model names.

Kastet commented on src/types/signable-token.ts:
```diff
@@ -0,0 +1,75 @@
+  type: 'ETH';
+}
+
+export type AnyToken = ETHToken | ERC721Token | ERC20Token;
```
Rename suggestion `AnyToken` -> `Token`

zatlite replied:
would clash with API.Token

Kastet commented on src/types/signable-token.ts:
```diff
@@ -0,0 +1,75 @@
+}
+
+// Token structure required by API
+export type AnySignableToken =
```
Rename suggestion `AnySignableToken` -> `APIToken`

zatlite replied:
removed [49b44f6](https://github.com/immutable/imx-core-sdk/pull/149/commits/49b44f65339265b89707c7f295157fa814e1dd78)

zatlite committed:
refactor: as per code review comments

zatlite committed:
refactor: remove SignableToken types

zatlite committed:
refactor(ERC20Deposit): remove unused field symbol

[DX-1184]

Kastet commented on src/types/signable-token.ts:
```diff
@@ -0,0 +1,48 @@
+export interface ERC20Token {
+  type: 'ERC20';
+  tokenAddress: string;
+  decimals: number;
```
decimals will most likely go as part of https://immutable.atlassian.net/browse/DX-1184

CodeSchwert replied:
Yes, it will be inferred from the token contract now for sending back to the API, but since undecimalised amounts are now used, `decimals` won't be required anymore.

zatlite replied:
I've tested ERC20 decimals are not required in get signable endpoints for transfer/order/withdrawal endpoints. 
Removed decimals here: [064964e](https://github.com/immutable/imx-core-sdk/pull/149/commits/064964ea75ebb2336539c401fb697a48dbd14fea)

zatlite commented on src/types/custom-interfaces.ts:
```diff
@@ -0,0 +1,64 @@
+     * @type {string}
+     */
+    amount: string;
+    token: AnyToken;
```
This validation currently exists in the order api. 

```go
// Assert this is an order
if signableData.SellAsset.IsFungible() && signableData.BuyAsset.IsNonFungible() {
	return imxerrors.NewUser("invalid buy and sell asset type for an order")
}

```
Should we update sell token to only accept ETH/ERC20 ? as well as Buy token to `token: ERC721Token`
```suggestion
    token: ETHToken | ERC20Token;
```

@Kastet

zatlite replied:
Update: Trading team says the above restriction will be removed in the future so I guess we can keep this code as is.

zatlite committed:
refactor: simplify request interfaces

- Consolidate deposit/withdrawal interfaces with TokenAmount
- Simplify order/transfer request interfaces
- Remove decimals from ERC20 GetSignableToken requests [DX-1184]
- Remove token_id from ERC20 encode asset api call

zatlite committed:
fix: optional chaining error response for 404 check

Kastet commented on src/types/signable-token.ts:
```diff
@@ -0,0 +1,55 @@
+import { SignableToken } from '../api';
+
+export interface ERC20Token {
```
Nice work @zatlite ! This is so much cleaner now. One suggestion is to rename the file as `signable-token.ts` doesn't seem to make sense anymore. Maybe `tokens.ts`?

zatlite replied:
Good idea. [fc419f7](https://github.com/immutable/imx-core-sdk/pull/149/commits/fc419f7bcaa9ab23ddeef88c47a8f7389b7f90da)

Kastet reviewed (approved):
Huge improvement 👌
Left a suggestion to rename files with token interfaces.

zatlite committed:
refactor: renamed file signable-token to tokens

Kastet commented:
@zatlite please also update https://github.com/immutable/imx-core-sdk-tester/pull/2

matthewmuscat commented on src/types/tokens.ts:
```diff
@@ -0,0 +1,55 @@
+  amount: string;
+}
+
+export type AnyToken = ETHToken | ERC721Token | ERC20Token;
```
It is unfortunate we cannot use `Token` here. I don't love `AnyToken` but I cannot come up with a better recommendation.

CodeSchwert replied:
Maybe it could be called `Asset`?

matthewmuscat replied:
Will clash with the existing `Assets` won't it?

matthewmuscat reviewed (approved):
LGTM

matthewmuscat commented on src/types/tokens.ts:
```diff
@@ -0,0 +1,55 @@
+
+export interface ETHAmount extends ETHToken {
+  amount: string;
+}
```
Can we add a newline here.

zatlite replied:
Emacs kept deleting end of file line. I had to turn off the commit hooks to get this commit [8d42a84](https://github.com/immutable/imx-core-sdk/pull/149/commits/8d42a8478a0aec8c6b527dff0ccdd66c8840cf91)

matthewmuscat commented on src/types/tokens.ts:
```diff
@@ -0,0 +1,55 @@
+        },
+      };
+  }
+}
```
End of file new line too please.

zatlite committed:
chore: add new lines
